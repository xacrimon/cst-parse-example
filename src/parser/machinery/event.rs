use super::{kind::SyntaxKind, span::Span};
use crate::T;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Event {
    Enter {
        kind: SyntaxKind,
        preceded_by: usize,
    },
    Exit,
    Token {
        kind: SyntaxKind,
        span: Span,
    },
}

impl Event {
    pub fn tombstone() -> Self {
        Self::Enter {
            kind: T![tombstone],
            preceded_by: 0,
        }
    }

    pub fn is_tombstone(self) -> bool {
        matches!(
            self,
            Self::Enter {
                kind: T![tombstone],
                preceded_by: 0,
            },
        )
    }
}

impl Default for Event {
    fn default() -> Self {
        Self::tombstone()
    }
}
