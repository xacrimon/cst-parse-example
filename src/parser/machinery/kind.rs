use std::fmt::{self, Display};

use logos::{Lexer, Logos};

#[allow(clippy::manual_non_exhaustive)]
#[derive(Logos, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Copy)]
#[repr(u16)]
pub enum SyntaxKind {
    // Miscellaneous
    #[error]
    Invalid = 0,

    Tombstone,

    Eof,
    Root,
    BreakStmt,
    ReturnStmt,
    BlockStmt,
    WhileStmt,
    RepeatStmt,
    StmtList,
    IfStmt,
    ElseChain,
    ForNumStmt,
    ForGenStmt,
    FuncStmt,
    FuncArgs,
    SimpleExpr,
    Expr,
    VarArgExpr,
    BinOp,
    FuncCall,
    Index,
    ExprList,
    DeclStmt,
    DeclTarget,
    FuncExpr,
    PrefixOp,
    TableExpr,
    TableArrayElem,
    TableMapElem,
    TableGenericElem,
    AssignStmt,
    LiteralExpr,

    #[regex(r"[ \n\t\f\r]+", logos::skip)]
    Whitespace,

    #[regex("--", skip_comment)]
    Comment,

    // Operators
    #[token("+")]
    Plus,

    #[token("-")]
    Minus,

    #[token("*")]
    Star,

    #[token("/")]
    Slash,

    #[token("%")]
    Percent,

    #[token("^")]
    Caret,

    #[token("#")]
    Hash,

    #[token("&")]
    Ampersand,

    #[token("|")]
    Pipe,

    #[token("~")]
    Tilde,

    #[token("<<")]
    DLAngle,

    #[token(">>")]
    DRAngle,

    #[token("==")]
    Eq,

    #[token("~=")]
    NotEq,

    #[token("<=")]
    LEq,

    #[token(">=")]
    GEq,

    #[token("<")]
    LAngle,

    #[token(">")]
    RAngle,

    #[token("=")]
    Assign,

    #[token("//")]
    DSlash,

    #[token(".")]
    Dot,

    #[token("..")]
    DDot,

    // Keywords
    #[token("local")]
    Local,

    #[token("function")]
    Function,

    #[token("end")]
    End,

    #[token("in")]
    In,

    #[token("then")]
    Then,

    #[token("break")]
    Break,

    #[token("for")]
    For,

    #[token("do")]
    Do,

    #[token("until")]
    Until,

    #[token("else")]
    Else,

    #[token("while")]
    While,

    #[token("elseif")]
    ElseIf,

    #[token("if")]
    If,

    #[token("repeat")]
    Repeat,

    #[token("return")]
    Return,

    #[token("not")]
    Not,

    #[token("or")]
    Or,

    #[token("and")]
    And,

    #[token("<const>")]
    Const,

    #[token("<close>")]
    Close,

    // Literals
    #[token("nil")]
    Nil,

    #[token("true")]
    True,

    #[token("false")]
    False,

    #[regex(r#""(\\[\\"]|[^"])*""#)]
    #[regex(r#"'(\\[\\']|[^'])*'"#)]
    String,

    #[regex(r"\[=*\[", long_string)]
    LongString,

    #[regex(r"[0-9]+", priority = 2)]
    Int,

    #[regex(r"0x[0-9a-fA-F]+")]
    HexInt,

    #[regex(r"[0-9]+(\.[0-9]+)?([eE][+-]?[0-9]+)?")]
    Float,

    #[regex(r"0x[0-9a-fA-F]*\.[0-9a-fA-F]+([pP][+-][0-9a-fA-F]+)?")]
    HexFloat,

    // Grouping
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*", priority = 3)]
    Ident,

    #[token("(")]
    LParen,

    #[token(")")]
    RParen,

    #[token("{")]
    LCurly,

    #[token("}")]
    RCurly,

    #[token("[")]
    LBracket,

    #[token("]")]
    RBracket,

    #[token(":")]
    Colon,

    #[token("::")]
    DColon,

    #[token(",")]
    Comma,

    #[token("...")]
    TDot,

    #[token(";")]
    Semicolon,

    #[doc(hidden)]
    #[allow(clippy::upper_case_acronyms)]
    __LAST,
}

impl SyntaxKind {
    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::Whitespace | SyntaxKind::Comment)
    }
}

fn long_string(lexer: &mut Lexer<SyntaxKind>) {
    let delim_len = lexer.slice().len();
    let rem = lexer.remainder();

    for (i, _) in rem.char_indices() {
        if is_long_delimiter(&rem[i..i + delim_len], ']') {
            lexer.bump(i + delim_len);
            return;
        }
    }

    unreachable!()
}

fn skip_comment(lexer: &mut Lexer<SyntaxKind>) -> logos::Skip {
    let rem = lexer.remainder();

    if let Some(delim_len) = starts_with_long_delimiter(rem, '[') {
        lexer.bump(delim_len);
        skip_long_comment(lexer, delim_len);
        logos::Skip
    } else {
        for (i, _) in rem.char_indices() {
            let curr = &rem[i..];
            if curr.starts_with("\r\n") {
                lexer.bump(i - 1);
                return logos::Skip;
            }

            if curr.starts_with('\n') {
                lexer.bump(i);
                return logos::Skip;
            }
        }

        unreachable!();
    }
}

fn skip_long_comment(lexer: &mut Lexer<SyntaxKind>, delim_len: usize) {
    let rem = lexer.remainder();

    for (i, _) in rem.char_indices() {
        if is_long_delimiter(&rem[i..i + delim_len], ']') {
            lexer.bump(i + delim_len);
            return;
        }
    }

    unreachable!()
}

fn starts_with_long_delimiter(slice: &str, delim: char) -> Option<usize> {
    if !slice.starts_with("[[") && !slice.starts_with("[=]") {
        return None;
    }

    for (i, _) in slice.char_indices() {
        if is_long_delimiter(&slice[..i], delim) {
            return Some(i);
        }
    }

    None
}

fn is_long_delimiter(slice: &str, delim: char) -> bool {
    if slice.len() < 2 || !slice.starts_with(delim) || !slice.ends_with(delim) {
        return false;
    }

    slice.chars().filter(|c| *c == '=').count() + 2 == slice.len()
}

#[macro_export]
macro_rules! T {
    [invalid] => { $crate::parser::machinery::kind::SyntaxKind::Invalid };
    [tombstone] => { $crate::parser::machinery::kind::SyntaxKind::Tombstone };
    [eof] => { $crate::parser::machinery::kind::SyntaxKind::Eof };
    [root] => { $crate::parser::machinery::kind::SyntaxKind::Root };
    [break_stmt] => { $crate::parser::machinery::kind::SyntaxKind::BreakStmt };
    [return_stmt] => { $crate::parser::machinery::kind::SyntaxKind::ReturnStmt };
    [block_stmt] => { $crate::parser::machinery::kind::SyntaxKind::BlockStmt };
    [while_stmt] => { $crate::parser::machinery::kind::SyntaxKind::WhileStmt };
    [repeat_stmt] => { $crate::parser::machinery::kind::SyntaxKind::RepeatStmt };
    [stmt_list] => { $crate::parser::machinery::kind::SyntaxKind::StmtList };
    [if_stmt] => { $crate::parser::machinery::kind::SyntaxKind::IfStmt };
    [else_chain] => { $crate::parser::machinery::kind::SyntaxKind::ElseChain };
    [for_num_stmt] => { $crate::parser::machinery::kind::SyntaxKind::ForNumStmt };
    [for_gen_stmt] => { $crate::parser::machinery::kind::SyntaxKind::ForGenStmt };
    [func_stmt] => { $crate::parser::machinery::kind::SyntaxKind::FuncStmt };
    [func_args] => { $crate::parser::machinery::kind::SyntaxKind::FuncArgs };
    [simple_expr] => { $crate::parser::machinery::kind::SyntaxKind::SimpleExpr };
    [expr] => { $crate::parser::machinery::kind::SyntaxKind::Expr };
    [vararg_expr] => { $crate::parser::machinery::kind::SyntaxKind::VarArgExpr };
    [bin_op] => { $crate::parser::machinery::kind::SyntaxKind::BinOp };
    [func_call] => { $crate::parser::machinery::kind::SyntaxKind::FuncCall };
    [index] => { $crate::parser::machinery::kind::SyntaxKind::Index };
    [expr_list] => { $crate::parser::machinery::kind::SyntaxKind::ExprList };
    [decl_stmt] => { $crate::parser::machinery::kind::SyntaxKind::DeclStmt };
    [decl_target] => { $crate::parser::machinery::kind::SyntaxKind::DeclTarget };
    [func_expr] => { $crate::parser::machinery::kind::SyntaxKind::FuncExpr };
    [prefix_op] => { $crate::parser::machinery::kind::SyntaxKind::PrefixOp };
    [table_expr] => { $crate::parser::machinery::kind::SyntaxKind::TableExpr };
    [table_array_elem] => { $crate::parser::machinery::kind::SyntaxKind::TableArrayElem };
    [table_map_elem] => { $crate::parser::machinery::kind::SyntaxKind::TableMapElem };
    [table_generic_elem] => { $crate::parser::machinery::kind::SyntaxKind::TableGenericElem };
    [assign_stmt] => { $crate::parser::machinery::kind::SyntaxKind::AssignStmt };
    [literal_expr] => { $crate::parser::machinery::kind::SyntaxKind::LiteralExpr };
    [ident] => { $crate::parser::machinery::kind::SyntaxKind::Ident };
    [+] => { $crate::parser::machinery::kind::SyntaxKind::Plus };
    [-] => { $crate::parser::machinery::kind::SyntaxKind::Minus };
    [*] => { $crate::parser::machinery::kind::SyntaxKind::Star };
    [/] => { $crate::parser::machinery::kind::SyntaxKind::Slash };
    [%] => { $crate::parser::machinery::kind::SyntaxKind::Percent };
    [^] => { $crate::parser::machinery::kind::SyntaxKind::Caret };
    [#] => { $crate::parser::machinery::kind::SyntaxKind::Hash };
    [&] => { $crate::parser::machinery::kind::SyntaxKind::Ampersand };
    [|] => { $crate::parser::machinery::kind::SyntaxKind::Pipe };
    [~] => { $crate::parser::machinery::kind::SyntaxKind::Tilde };
    [<<] => { $crate::parser::machinery::kind::SyntaxKind::DLAngle };
    [>>] => { $crate::parser::machinery::kind::SyntaxKind::DRAngle };
    [==] => { $crate::parser::machinery::kind::SyntaxKind::Eq };
    [~=] => { $crate::parser::machinery::kind::SyntaxKind::NotEq };
    [<=] => { $crate::parser::machinery::kind::SyntaxKind::LEq };
    [>=] => { $crate::parser::machinery::kind::SyntaxKind::GEq };
    [<] => { $crate::parser::machinery::kind::SyntaxKind::LAngle };
    [>] => { $crate::parser::machinery::kind::SyntaxKind::RAngle };
    [=] => { $crate::parser::machinery::kind::SyntaxKind::Assign };
    [D/] => { $crate::parser::machinery::kind::SyntaxKind::DSlash };
    [local] => { $crate::parser::machinery::kind::SyntaxKind::Local };
    [function] => { $crate::parser::machinery::kind::SyntaxKind::Function };
    [end] => { $crate::parser::machinery::kind::SyntaxKind::End };
    [in] => { $crate::parser::machinery::kind::SyntaxKind::In };
    [then] => { $crate::parser::machinery::kind::SyntaxKind::Then };
    [break] => { $crate::parser::machinery::kind::SyntaxKind::Break };
    [for] => { $crate::parser::machinery::kind::SyntaxKind::For };
    [do] => { $crate::parser::machinery::kind::SyntaxKind::Do };
    [until] => { $crate::parser::machinery::kind::SyntaxKind::Until };
    [else] => { $crate::parser::machinery::kind::SyntaxKind::Else };
    [while] => { $crate::parser::machinery::kind::SyntaxKind::While };
    [elseif] => { $crate::parser::machinery::kind::SyntaxKind::ElseIf };
    [if] => { $crate::parser::machinery::kind::SyntaxKind::If };
    [repeat] => { $crate::parser::machinery::kind::SyntaxKind::Repeat };
    [return] => { $crate::parser::machinery::kind::SyntaxKind::Return };
    [not] => { $crate::parser::machinery::kind::SyntaxKind::Not };
    [or] => { $crate::parser::machinery::kind::SyntaxKind::Or };
    [and] => { $crate::parser::machinery::kind::SyntaxKind::And };
    [const] => { $crate::parser::machinery::kind::SyntaxKind::Const };
    [close] => { $crate::parser::machinery::kind::SyntaxKind::Close };
    [nil] => { $crate::parser::machinery::kind::SyntaxKind::Nil };
    [true] => { $crate::parser::machinery::kind::SyntaxKind::True };
    [false] => { $crate::parser::machinery::kind::SyntaxKind::False };
    [string] => { $crate::parser::machinery::kind::SyntaxKind::String };
    [long_string] => { $crate::parser::machinery::kind::SyntaxKind::LongString };
    [int] => { $crate::parser::machinery::kind::SyntaxKind::Int };
    [hex_int] => { $crate::parser::machinery::kind::SyntaxKind::HexInt };
    [float] => { $crate::parser::machinery::kind::SyntaxKind::Float };
    [hex_float] => { $crate::parser::machinery::kind::SyntaxKind::HexFloat };
    ['('] => { $crate::parser::machinery::kind::SyntaxKind::LParen };
    [')'] => { $crate::parser::machinery::kind::SyntaxKind::RParen };
    ['{'] => { $crate::parser::machinery::kind::SyntaxKind::LCurly };
    ['}'] => { $crate::parser::machinery::kind::SyntaxKind::RCurly };
    ['['] => { $crate::parser::machinery::kind::SyntaxKind::LBracket };
    [']'] => { $crate::parser::machinery::kind::SyntaxKind::RBracket };
    [:] => { $crate::parser::machinery::kind::SyntaxKind::Colon };
    [::] => { $crate::parser::machinery::kind::SyntaxKind::DColon };
    [,] => { $crate::parser::machinery::kind::SyntaxKind::Comma };
    [.] => { $crate::parser::machinery::kind::SyntaxKind::Dot };
    [..] => { $crate::parser::machinery::kind::SyntaxKind::DDot };
    [...] => { $crate::parser::machinery::kind::SyntaxKind::TDot };
    [;] => { $crate::parser::machinery::kind::SyntaxKind::Semicolon };
    [__LAST] => { $crate::parser::machinery::kind::SyntaxKind::__LAST };
}

impl Display for SyntaxKind {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                T![invalid] => "INVALID",
                T![eof] => "EOF",
                T![root] => "ROOT",
                T![ident] => "IDENTIFIER",
                T![+] => "PLUS",
                T![-] => "MINUS",
                T![*] => "STAR",
                T![/] => "SLASH",
                T![%] => "PERCENT",
                T![^] => "CARET",
                T![#] => "HASH",
                T![&] => "AMPERSAND",
                T![|] => "PIPE",
                T![~] => "TILDE",
                T![<<] => "DLANGLE",
                T![>>] => "DRANGLE",
                T![==] => "EQ",
                T![~=] => "NOT_EQ",
                T![<=] => "LEQ",
                T![>=] => "GEQ",
                T![<] => "LANGLE",
                T![>] => "RANGLE",
                T![=] => "ASSIGN",
                T![D/] => "DSLASH",
                T![local] => "LOCAL",
                T![function] => "FUNCTION",
                T![end] => "END",
                T![in] => "IN",
                T![then] => "THEN",
                T![break] => "BREAK",
                T![for] => "FOR",
                T![do] => "DO",
                T![until] => "UNTIL",
                T![else] => "ELSE",
                T![while] => "WHILE",
                T![elseif] => "ELSEIF",
                T![if] => "IF",
                T![repeat] => "REPEAT",
                T![return] => "RETURN",
                T![not] => "NOT",
                T![or] => "OR",
                T![and] => "AND",
                T![const] => "CONST",
                T![close] => "CLOSE",
                T![nil] => "NIL",
                T![true] => "TRUE",
                T![false] => "FALSE",
                T![string] => "STRING",
                T![long_string] => "LONG_STRING",
                T![int] => "INT",
                T![hex_int] => "HEX_INT",
                T![float] => "FLOAT",
                T![hex_float] => "HEX_FLOAT",
                T!['('] => "RPAREN",
                T![')'] => "LPAREN",
                T!['{'] => "RCURLY",
                T!['}'] => "LCURLY",
                T!['['] => "RBRACKET",
                T![']'] => "LBRACKET",
                T![:] => "COLON",
                T![::] => "DCOLON",
                T![,] => "COMMA",
                T![.] => "DOT",
                T![..] => "DDOT",
                T![...] => "TDOT",
                _ => unreachable!(),
            }
        )
    }
}
