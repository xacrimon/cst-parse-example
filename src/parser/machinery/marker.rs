use super::{event::Event, kind::SyntaxKind, state::State};
use crate::T;

pub struct Marker {
    position: usize,
}

impl Marker {
    pub fn new(position: usize) -> Self {
        Self { position }
    }

    pub fn complete(self, state: &mut State, kind: SyntaxKind) -> CompletedMarker {
        let event_at_pos = &mut state.events()[self.position];
        debug_assert_eq!(*event_at_pos, Event::tombstone());

        *event_at_pos = Event::Enter {
            kind,
            preceded_by: 0,
        };

        state.events().push(Event::Exit);
        CompletedMarker {
            position: self.position,
            kind,
        }
    }

    pub fn abandon(self, state: &mut State) {
        match &mut state.events()[self.position] {
            Event::Enter {
                kind,
                preceded_by: 0,
            } => {
                *kind = T![tombstone];
            },

            _ => unreachable!(),
        }

        debug_assert_eq!(state.events()[self.position], Event::tombstone());
        if self.position == state.events().len() - 1 {
            state.events().pop();
        }
    }
}

#[derive(Debug)]
pub struct CompletedMarker {
    position: usize,
    kind: SyntaxKind,
}

impl CompletedMarker {
    pub fn precede(self, state: &mut State) -> Marker {
        let marker = state.start();

        if let Event::Enter { preceded_by, .. } = &mut state.events()[self.position] {
            *preceded_by = marker.position - self.position;
        } else {
            unreachable!();
        }

        marker
    }
}
