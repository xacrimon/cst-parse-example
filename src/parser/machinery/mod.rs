pub mod binding_power;
pub mod classifiers;
pub mod event;
pub mod kind;
pub mod marker;
pub mod sink;
pub mod span;
pub mod state;
