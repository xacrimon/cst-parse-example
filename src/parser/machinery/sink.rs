use std::mem;

use cstree::{GreenNode, GreenNodeBuilder, NodeCache};

use super::{event::Event, kind::SyntaxKind, span::Span};
use crate::T;

pub struct Sink<'cache, 'source> {
    builder: GreenNodeBuilder<'cache, 'static>,
    tokens: &'source [(SyntaxKind, Span)],
    cursor: usize,
    events: Vec<Event>,
    source: &'source str,
}

impl<'cache, 'source> Sink<'cache, 'source> {
    pub fn new(
        cache: &'cache mut NodeCache<'static>,
        tokens: &'source [(SyntaxKind, Span)],
        events: Vec<Event>,
        source: &'source str,
    ) -> Self {
        Self {
            builder: GreenNodeBuilder::with_cache(cache),
            tokens,
            cursor: 0,
            events,
            source,
        }
    }

    fn token(&mut self, kind: SyntaxKind, text: &str) {
        self.cursor += 1;
        self.builder.token(kind.into(), text);
    }

    pub fn finish(mut self) -> GreenNode {
        let mut preceded_nodes = Vec::new();
        for idx in 0..self.events.len() {
            match mem::take(&mut self.events[idx]) {
                // Ignore tombstone events
                event @ Event::Enter { .. } if event.is_tombstone() => {},

                Event::Enter { kind, preceded_by } => {
                    preceded_nodes.push(kind);

                    let (mut idx, mut preceded_by) = (idx, preceded_by);
                    while preceded_by > 0 {
                        idx += preceded_by;

                        preceded_by = match mem::take(&mut self.events[idx]) {
                            Event::Enter { kind, preceded_by } => {
                                if kind != T![tombstone] {
                                    preceded_nodes.push(kind);
                                }

                                preceded_by
                            },

                            _ => unreachable!(),
                        }
                    }

                    for kind in preceded_nodes.drain(..).rev() {
                        self.builder.start_node(kind.into());
                    }
                },

                Event::Exit => {
                    self.builder.finish_node();
                },

                Event::Token { kind, span } => {
                    self.token(kind, &self.source[span]);
                },
            }
        }

        self.builder.finish().0
    }
}
