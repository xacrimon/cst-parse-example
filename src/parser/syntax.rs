use crate::{parser::machinery::kind::SyntaxKind, T};

impl From<SyntaxKind> for cstree::SyntaxKind {
    fn from(token: SyntaxKind) -> Self {
        Self(token as u16)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Lang {}

impl cstree::Language for Lang {
    type Kind = SyntaxKind;

    fn kind_from_raw(raw: cstree::SyntaxKind) -> Self::Kind {
        debug_assert!(raw.0 < T![__LAST] as u16);
        unsafe { std::mem::transmute(raw.0) }
    }

    fn kind_to_raw(kind: Self::Kind) -> cstree::SyntaxKind {
        kind.into()
    }
}

pub type SyntaxNode = cstree::SyntaxNode<Lang>;
pub type SyntaxToken = cstree::SyntaxToken<Lang>;
pub type SyntaxElement = cstree::NodeOrToken<SyntaxNode, SyntaxToken>;

macro_rules! ast_node {
    ($ast:ident, $kind:expr) => {
        #[derive(PartialEq, Eq, Hash)]
        #[repr(transparent)]
        struct $ast(SyntaxNode);
        impl $ast {
            #[allow(unused)]
            fn cast(node: SyntaxNode) -> Option<Self> {
                if node.kind() == $kind {
                    Some(Self(node))
                } else {
                    None
                }
            }
        }
    };
}

ast_node!(Root, T![root]);
ast_node!(DeclStmt, T![decl_stmt]);
ast_node!(DeclTarget, T![decl_target]);
ast_node!(LiteralExpr, T![literal_expr]);
ast_node!(AssignStmt, T![assign_stmt]);
ast_node!(Ident, T![ident]);
ast_node!(PrefixOp, T![prefix_op]);
ast_node!(BinaryOp, T![bin_op]);
ast_node!(ExprList, T![expr_list]);
ast_node!(FuncCall, T![func_call]);
ast_node!(Func, T![function]);
ast_node!(StmtList, T![stmt_list]);
