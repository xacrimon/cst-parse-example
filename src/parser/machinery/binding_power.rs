use super::kind::SyntaxKind;
use crate::T;

pub const INDEX_BINDING_POWER: i32 = 22;
pub const CALL_BINDING_POWER: i32 = 22;

pub fn prefix_binding_power(op: SyntaxKind) -> ((), i32) {
    match op {
        T![not] | T![+] | T![-] | T![#] | T![~] => ((), 21),
        _ => unreachable!(),
    }
}

pub fn infix_binding_power(op: SyntaxKind) -> Option<(i32, i32)> {
    Some(match op {
        T![or] => (1, 2),
        T![and] => (3, 4),
        T![<] | T![>] | T![<=] | T![>=] | T![~=] | T![==] => (5, 6),
        T![|] => (7, 8),
        T![~] => (9, 10),
        T![&] => (11, 12),
        T![<<] | T![>>] => (13, 14),
        T![..] => (16, 15),
        T![+] | T![-] => (17, 18),
        T![*] | T![/] | T![D/] | T![%] => (19, 20),
        T![^] => (22, 21),
        T![.] | T![:] => (24, 23),
        _ => return None,
    })
}
