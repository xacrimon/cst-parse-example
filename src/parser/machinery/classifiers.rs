use super::kind::SyntaxKind;
use crate::T;

pub fn token_is_literal(token: SyntaxKind) -> bool {
    matches!(
        token,
        T![nil]
            | T![false]
            | T![true]
            | T![int]
            | T![hex_int]
            | T![float]
            | T![hex_float]
            | T![string]
            | T![long_string]
    )
}

pub fn token_is_expr_start(token: SyntaxKind) -> bool {
    token == T![ident]
        || token == T!['(']
        || token_is_literal(token)
        || token_is_unary_op(token)
        || token == T!['{']
        || token == T![function]
        || token == T![...]
}

pub fn token_is_unary_op(token: SyntaxKind) -> bool {
    matches!(token, T![not] | T![+] | T![-] | T![#] | T![~])
}

pub fn token_is_binary_op(token: SyntaxKind) -> bool {
    matches!(
        token,
        T![or]
            | T![and]
            | T![+]
            | T![-]
            | T![*]
            | T![/]
            | T![D/]
            | T![^]
            | T![%]
            | T![&]
            | T![|]
            | T![<<]
            | T![>>]
            | T![==]
            | T![~]
            | T![~=]
            | T![<=]
            | T![>=]
            | T![>]
            | T![<]
            | T![.]
            | T![:]
            | T![..]
    )
}
