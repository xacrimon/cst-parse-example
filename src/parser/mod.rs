mod assign;
mod control;
mod expr;
mod function;
mod item;
pub mod machinery;
mod simple_expr;
mod stmt;
mod syntax;
mod table;

use std::ops::{Deref, DerefMut};

use cstree::NodeCache;
use machinery::{span::Span, state::State};
use syntax::SyntaxNode;

use crate::T;

struct Parser<'cache, 'source> {
    state: State<'cache, 'source>,
}

impl<'cache, 'source> Parser<'cache, 'source> {
    fn new(cache: &'cache mut NodeCache<'static>, source: &'source str) -> Self {
        Self {
            state: State::new(cache, source),
        }
    }

    fn root(&mut self) {
        let marker = self.start();
        self.r_items();
        marker.complete(self, T![root]);
    }

    fn run(mut self) -> (SyntaxNode, Vec<ariadne::Report<Span>>) {
        self.root();
        let (root, reports) = self.state.finish();
        (SyntaxNode::new_root(root), reports)
    }
}

impl<'cache, 'source> Deref for Parser<'cache, 'source> {
    type Target = State<'cache, 'source>;

    fn deref(&self) -> &Self::Target {
        &self.state
    }
}

impl<'cache, 'source> DerefMut for Parser<'cache, 'source> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.state
    }
}

pub fn parse(
    cache: &mut NodeCache<'static>,
    source: &str,
) -> (SyntaxNode, Vec<ariadne::Report<Span>>) {
    Parser::new(cache, source).run()
}

#[cfg(test)]
mod tests {
    use std::fs;

    use cstree::NodeCache;
    use insta::assert_snapshot;
    use paste::paste;

    use super::{parse, syntax::SyntaxNode};

    fn syntax_tree_debug(cache: &NodeCache<'static>, node: &SyntaxNode) -> String {
        node.debug(cache.interner(), true)
    }

    macro_rules! parse_and_verify {
        ($name:ident, $path:literal) => {
            paste! {
                #[test]
                fn [<parse_and_verify_ $name>]() {
                    let mut cache = NodeCache::new();
                    let source = fs::read_to_string($path).unwrap();
                    let (syntax_tree, reports) = parse(&mut cache, &source);
                    let syntax_tree_debug = syntax_tree_debug(&cache, &syntax_tree);
                    assert!(reports.is_empty());
                    assert_snapshot!(syntax_tree_debug);
                }
            }
        };
    }

    parse_and_verify!(function, "test-files/function.lua");
    parse_and_verify!(op_prec, "test-files/op_prec.lua");
    parse_and_verify!(if, "test-files/if.lua");
    parse_and_verify!(declare, "test-files/declare.lua");
    parse_and_verify!(literal, "test-files/literal.lua");
    parse_and_verify!(comment, "test-files/comment.lua");
    parse_and_verify!(mixed, "test-files/mixed.lua");
}
