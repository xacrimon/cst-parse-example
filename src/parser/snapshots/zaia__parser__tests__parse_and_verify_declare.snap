---
source: src/parser/mod.rs
assertion_line: 98
expression: syntax_tree_debug

---
Root@0..175
  AssignStmt@0..3
    Ident@0..1
      Ident@0..1 "x"
    Assign@1..2 "="
    LiteralExpr@2..3
      Int@2..3 "5"
  DeclStmt@3..12
    Local@3..8 "local"
    DeclTarget@8..9
      Ident@8..9 "y"
    Assign@9..10 "="
    LiteralExpr@10..12
      Int@10..12 "10"
  DeclStmt@12..22
    Local@12..17 "local"
    DeclTarget@17..19
      Ident@17..19 "aq"
    Comma@19..20 ","
    DeclTarget@20..22
      Ident@20..22 "bf"
  DeclStmt@22..38
    Local@22..27 "local"
    DeclTarget@27..35
      Ident@27..28 "z"
      Const@28..35 "<const>"
    Assign@35..36 "="
    LiteralExpr@36..38
      Int@36..38 "15"
  DeclStmt@38..57
    Local@38..43 "local"
    DeclTarget@43..51
      Ident@43..44 "w"
      Close@44..51 "<close>"
    Assign@51..52 "="
    LiteralExpr@52..57
      String@52..57 "\"abs\""
  AssignStmt@57..75
    Ident@57..58
      Ident@57..58 "i"
    Comma@58..59 ","
    Ident@59..60
      Ident@59..60 "k"
    Assign@60..61 "="
    ExprList@61..75
      FuncCall@61..66
        Ident@61..64
          Ident@61..64 "get"
        FuncArgs@64..66
          LParen@64..65 "("
          RParen@65..66 ")"
      Comma@66..67 ","
      LiteralExpr@67..75
        String@67..75 "\"other1\""
  DeclStmt@75..99
    Local@75..80 "local"
    DeclTarget@80..81
      Ident@80..81 "r"
    Comma@81..82 ","
    DeclTarget@82..83
      Ident@82..83 "g"
    Assign@83..84 "="
    ExprList@84..99
      FuncCall@84..90
        Ident@84..88
          Ident@84..88 "getf"
        FuncArgs@88..90
          LParen@88..89 "("
          RParen@89..90 ")"
      Comma@90..91 ","
      LiteralExpr@91..99
        String@91..99 "\"other2\""
  DeclStmt@99..137
    Local@99..104 "local"
    DeclTarget@104..112
      Ident@104..105 "f"
      Const@105..112 "<const>"
    Comma@112..113 ","
    DeclTarget@113..121
      Ident@113..114 "v"
      Const@114..121 "<const>"
    Assign@121..122 "="
    ExprList@122..137
      FuncCall@122..128
        Ident@122..126
          Ident@122..126 "getw"
        FuncArgs@126..128
          LParen@126..127 "("
          RParen@127..128 ")"
      Comma@128..129 ","
      LiteralExpr@129..137
        String@129..137 "\"other3\""
  DeclStmt@137..175
    Local@137..142 "local"
    DeclTarget@142..150
      Ident@142..143 "p"
      Close@143..150 "<close>"
    Comma@150..151 ","
    DeclTarget@151..159
      Ident@151..152 "o"
      Close@152..159 "<close>"
    Assign@159..160 "="
    ExprList@160..175
      FuncCall@160..166
        Ident@160..164
          Ident@160..164 "getq"
        FuncArgs@164..166
          LParen@164..165 "("
          RParen@165..166 ")"
      Comma@166..167 ","
      LiteralExpr@167..175
        String@167..175 "\"other4\""

